include/config/config_interface.h

```diff
@@ -0,0 +1,113 @@
+#pragma once
+
+#include <iostream>
+#include <optional>
+#include <span>
+#include <variant>
+
+#include <spdlog/spdlog.h>
+
+#include "config/config_exception.h"
+#include "config/config_source_interface.h"
+#include "config/config_specification.h"
+#include "config/source/command_line_arguments.h"
+#include "config/source/environment_variables.h"
+#include "config/source/yaml_file.h"
+#include "silo/common/cons_list.h"
+#include "silo/common/fmt_formatters.h"
+#include "silo/common/overloaded.h"
+
+namespace silo::config {
+
+/// For config structs (possibly containing config file paths).
+// We use a concept instead of virtual methods, because the context of
+// its usage (getConfig) is using a template anyways, due to the
+// different return types (RuntimeConfig vs.
+// PreprocessingConfig). Alternatively, making the constructor private
+// and instead creating a factory method would also be possible.
+template <typename C>
+concept Config = requires(C c, const C cc, const VerifiedConfigAttributes& config_source) {
+   { C::getConfigSpecification() } -> std::same_as<ConfigSpecification>;
+
+   /// Vector of config files that the user gave (or that is provided
+   /// by the type via its defaults) that should be loaded and shadowed
+   /// in the order of the vector.
+   { cc.getConfigPaths() } -> std::same_as<std::vector<std::filesystem::path>>;
+
+   /// Overwrite the fields of an instance of the target type; done
+   /// that way so that multiple kinds of config sources can shadow
+   /// each other's values by application in sequence. Does not throw
+   /// exceptions, except overwriteFrom can call SILO_PANIC when there
+   /// is an inconsistency (bug) between ConfigSpecification and
+   /// overwriteFrom implementation.
+   { c.overwriteFrom(config_source) } -> std::same_as<void>;
+
+   /// Validation / Sanity checks about the values of this config
+   { c.validate() } -> std::same_as<void>;
+};
+
+/// This function needs a reference to the (remaining) command line
+/// arguments to be parsed, thus the application gets a chance to take
+/// off some arguments first (like "api" or "preprocessing" in the
+/// current SILO version).
+///
+/// In case of error, returns the exit code that the caller should
+/// pass to exit(): 0 if the user gave --help, 1 in case of erroneous
+/// usage (the error is already printed in that case).
+template <Config C>
+std::variant<C, int32_t> getConfig(
+   std::span<const std::string> cmd,
+   const std::vector<std::string>& allow_list_for_env_vars
+) {
+   const auto config_specification = C::getConfigSpecification();
+   try {
+      auto cmd_source = CommandLineArguments{cmd}.verify(config_specification);
+      if (cmd_source.asks_for_help) {
+         std::cout << config_specification.helpText() << "\n" << std::flush;
+         return 0;
+      }
+      if (!cmd_source.positional_arguments.empty()) {
+         throw silo::config::ConfigException{"SILO does not expect positional arguments"};
```

Comment:
```suggestion
         throw silo::config::ConfigException{"SILO does not expect positional arguments, found {cmd_source.positional_arguments}"};
```
(in suitable non-pseudo code)